File: python/agentshield/runtime/enforcer.py
"""Runtime enforcer — IFC + tool policy with fail-closed / dev-warn modes."""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Mapping

from agentshield.mcp.firewall import FirewallDecision
from agentshield.runtime.audit_chain import AuditChain
from agentshield.runtime.config import RuntimeMode
from agentshield.runtime.context import ExecutionContext
from agentshield.runtime.policy_dsl import RuntimePolicy
from agentshield.runtime.quarantine import QuarantineStore


class EnforcementDecision(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    REQUIRE_APPROVAL = "require_approval"
    WARN = "warn"


@dataclass(frozen=True)
class EnforcementResult:
    decision: EnforcementDecision
    tool_name: str
    reason: str
    rule_id: str = ""
    elapsed_ms: float = 0.0
    labels_in_scope: dict[str, str] | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "decision": self.decision.value,
            "tool_name": self.tool_name,
            "reason": self.reason,
            "rule_id": self.rule_id,
            "elapsed_ms": self.elapsed_ms,
            "labels_in_scope": self.labels_in_scope,
        }

    @property
    def blocked(self) -> bool:
        return self.decision in (EnforcementDecision.DENY, EnforcementDecision.REQUIRE_APPROVAL)


@dataclass
class RuntimeEnforcer:
    """Evaluate tool calls against labels in scope and runtime policy."""

    policy: RuntimePolicy
    context: ExecutionContext = field(default_factory=ExecutionContext)
    quarantine: QuarantineStore = field(default_factory=QuarantineStore)
    audit: AuditChain | None = None

    def ingest_content(
        self,
        content: str,
        *,
        integrity: str = "untrusted",
        confidentiality: str = "public",
        source: str = "external",
    ) -> str:
        from agentshield.runtime.labels import Confidentiality, FlowLabels, Integrity

        labels = FlowLabels(
            integrity=Integrity.from_string(integrity),
            confidentiality=Confidentiality.from_string(confidentiality),
        )
        visible = content
        if self.policy.quarantine_untrusted and labels.integrity is Integrity.UNTRUSTED:
            visible, labels, ref = self.quarantine.ingest_if_untrusted(
                content, labels, source=source
            )
            if ref and self.audit:
                self.audit.append(
                    "quarantine_store",
                    {"ref_id": ref.ref_id, "source": source, "digest": ref.digest},
                )
        self.context.ingest(content, labels=labels, source=source)
        return visible

    def check_tool_call(self, tool_call: Mapping[str, Any]) -> EnforcementResult:
        start = time.perf_counter()
        name = str(tool_call.get("name") or tool_call.get("tool") or "").strip()
        if not name:
            return self._finalize(
                EnforcementResult(
                    EnforcementDecision.DENY,
                    "",
                    "empty_tool_name",
                    rule_id="runtime.empty",
                ),
                start,
            )

        # Blocked patterns first
        for pattern in self.policy.tool_policy.blocked_patterns:
            try:
                if re.search(pattern, name):
                    return self._finalize(
                        EnforcementResult(
                            EnforcementDecision.DENY,
                            name,
                            f"blocked_pattern:{pattern}",
                            rule_id="tool_policy.blocked",
                        ),
                        start,
                    )
            except re.error:
                if pattern and pattern in name:
                    return self._finalize(
                        EnforcementResult(
                            EnforcementDecision.DENY,
                            name,
                            f"blocked_pattern:{pattern}",
                            rule_id="tool_policy.blocked",
                        ),
                        start,
                    )

        # IFC rules before allowlist — untrusted scope blocks sensitive tools
        for idx, rule in enumerate(self.policy.ifc_rules):
            if not rule.matches(
                has_untrusted=self.context.has_untrusted,
                has_private=self.context.has_private,
            ):
                continue
            if name in rule.deny_tools or self._matches_sensitive(name, rule.deny_tools):
                return self._finalize(
                    EnforcementResult(
                        EnforcementDecision.DENY,
                        name,
                        rule.reason,
                        rule_id=f"ifc.deny.{idx}",
                    ),
                    start,
                )
            if name in rule.require_approval_tools:
                return self._finalize(
                    EnforcementResult(
                        EnforcementDecision.REQUIRE_APPROVAL,
                        name,
                        rule.reason,
                        rule_id=f"ifc.approval.{idx}",
                    ),
                    start,
                )

        if self.context.has_untrusted and name in set(self.policy.sensitive_tools):
            return self._finalize(
                EnforcementResult(
                    EnforcementDecision.DENY,
                    name,
                    "sensitive_tool_with_untrusted_in_scope",
                    rule_id="ifc.sensitive_default",
                ),
                start,
            )

        # Phase 1 tool allowlist (allow / approval / deny)
        fw = self.policy.tool_policy.allowlist().decide(name)
        if fw.decision is FirewallDecision.DENY:
            return self._finalize(
                EnforcementResult(
                    EnforcementDecision.DENY,
                    name,
                    fw.reason,
                    rule_id="tool_policy.deny",
                ),
                start,
            )
        if fw.decision is FirewallDecision.REQUIRE_APPROVAL:
            return self._finalize(
                EnforcementResult(
                    EnforcementDecision.REQUIRE_APPROVAL,
                    name,
                    fw.reason,
                    rule_id="tool_policy.approval",
                ),
                start,
            )

        return self._finalize(
            EnforcementResult(
                EnforcementDecision.ALLOW,
                name,
                "allowed",
                rule_id="runtime.allow",
            ),
            start,
        )

    def _matches_sensitive(self, tool_name: str, patterns: list[str]) -> bool:
        for pattern in patterns:
            if pattern == tool_name:
                return True
            try:
                if re.search(pattern, tool_name):
                    return True
            except re.error:
                continue
        return False

    def _finalize(self, result: EnforcementResult, start: float) -> EnforcementResult:
        elapsed_ms = (time.perf_counter() - start) * 1000.0
        labels = self.context.labels_in_scope.to_dict()
        decision = result.decision

        if (
            self.policy.mode is RuntimeMode.DEVELOPMENT
            and result.decision is EnforcementDecision.DENY
        ):
            decision = EnforcementDecision.WARN

        finalized = EnforcementResult(
            decision=decision,
            tool_name=result.tool_name,
            reason=result.reason,
            rule_id=result.rule_id,
            elapsed_ms=round(elapsed_ms, 3),
            labels_in_scope=labels,
        )

        if self.audit:
            self.audit.append(
                "tool_check",
                {
                    "tool": result.tool_name,
                    "decision": finalized.decision.value,
                    "reason": finalized.reason,
                    "rule_id": finalized.rule_id,
                    "elapsed_ms": finalized.elapsed_ms,
                    "labels": labels,
                },
            )

        self.context.record_tool_call(
            {"name": result.tool_name, "decision": finalized.to_dict()}
        )
        return finalized

    def evaluate_trace(self, trace: Mapping[str, Any]) -> list[EnforcementResult]:
        """Replay a trace: ingest content steps then evaluate tool calls."""
        self.context = ExecutionContext.from_trace(dict(trace))
        results: list[EnforcementResult] = []
        for step in trace.get("steps") or []:
            if step.get("type") == "content":
                self.ingest_content(
                    step.get("content") or "",
                    integrity=(step.get("labels") or {}).get("integrity", "untrusted"),
                    confidentiality=(step.get("labels") or {}).get(
                        "confidentiality", "public"
                    ),
                    source=step.get("source") or "external",
                )
            elif step.get("type") == "tool_call":
                call = step.get("call") or step
                results.append(self.check_tool_call(call))
        return results

File: server/services/runtime/enforcer.js
'use strict';

const { Integrity, Confidentiality, FlowLabels } = require('./labels');
const { AuditChain } = require('./auditChain');

const EnforcementDecision = {
  ALLOW: 'allow',
  DENY: 'deny',
  REQUIRE_APPROVAL: 'require_approval',
  WARN: 'warn'
};

class ExecutionContext {
  constructor() {
    this.labelsInScope = FlowLabels.trustedPublic();
    this.contentRecords = [];
    this.toolCalls = [];
  }

  ingest(content, labels = FlowLabels.trustedPublic(), source = 'user', ref = null) {
    this.labelsInScope = this.labelsInScope.merge(labels);
    this.contentRecords.push({
      ref: ref || `content_${this.contentRecords.length + 1}`,
      labels: labels.toDict(),
      source,
      byteLength: Buffer.byteLength(content, 'utf8')
    });
    return this.labelsInScope;
  }

  static fromTrace(trace) {
    const ctx = new ExecutionContext();
    for (const step of trace.steps || []) {
      if (step.type === 'content') {
        ctx.ingest(
          step.content || '',
          FlowLabels.fromDict(step.labels),
          step.source || 'unknown',
          step.ref
        );
      }
    }
    if (trace.labels_in_scope) {
      ctx.labelsInScope = FlowLabels.fromDict(trace.labels_in_scope);
    }
    return ctx;
  }

  get hasUntrusted() {
    return this.labelsInScope.integrity === Integrity.UNTRUSTED;
  }

  get hasPrivate() {
    return this.labelsInScope.confidentiality === Confidentiality.PRIVATE;
  }
}

function matchesIfcRule(rule, ctx) {
  const when = rule.when || {};
  for (const [key, expected] of Object.entries(when)) {
    const norm = String(expected).toLowerCase();
    if (key === 'integrity_in_scope' || key === 'integrity') {
      const actual = ctx.hasUntrusted ? 'untrusted' : 'trusted';
      if (actual !== norm) return false;
    } else if (key === 'confidentiality_in_scope' || key === 'confidentiality') {
      const actual = ctx.hasPrivate ? 'private' : 'public';
      if (actual !== norm) return false;
    } else {
      return false;
    }
  }
  return true;
}

function decideToolPolicy(toolPolicy, toolName) {
  const name = (toolName || '').trim();
  if (!name) return { decision: 'deny', reason: 'empty_tool_name' };
  for (const pattern of toolPolicy.blocked_patterns || []) {
    try {
      if (new RegExp(pattern).test(name)) {
        return { decision: 'deny', reason: `blocked_pattern:${pattern}` };
      }
    } catch {
      if (pattern && name.includes(pattern)) {
        return { decision: 'deny', reason: `blocked_pattern:${pattern}` };
      }
    }
  }
  const allowed = new Set(toolPolicy.allowed_tools || []);
  const approval = new Set(toolPolicy.approval_required || []);
  if (allowed.has(name)) return { decision: 'allow', reason: 'allowlisted' };
  if (approval.has(name)) return { decision: 'require_approval', reason: 'approval_required' };
  if (toolPolicy.fail_closed !== false) {
    return { decision: 'deny', reason: 'not_allowlisted' };
  }
  return { decision: 'require_approval', reason: 'fail_open_approval' };
}

function evaluateTrace(trace, runtimePolicy = {}) {
  const mode = (runtimePolicy.mode || 'production').toLowerCase();
  const toolPolicy = runtimePolicy.tool_policy || runtimePolicy;
  const ifcRules = runtimePolicy.ifc_rules || [];
  const sensitive = new Set(runtimePolicy.sensitive_tools || []);
  const ctx = ExecutionContext.fromTrace(trace);
  const results = [];

  for (const step of trace.steps || []) {
    if (step.type === 'content') {
      ctx.ingest(
        step.content || '',
        FlowLabels.fromDict(step.labels),
        step.source || 'external'
      );
      continue;
    }
    if (step.type !== 'tool_call') continue;

    const call = step.call || step;
    const name = String(call.name || call.tool || '').trim();
    const start = Date.now();
    let decision = EnforcementDecision.ALLOW;
    let reason = 'allowed';
    let ruleId = 'runtime.allow';

    const blocked = toolPolicy.blocked_patterns || [];
    for (const pattern of blocked) {
      try {
        if (new RegExp(pattern).test(name)) {
          decision = EnforcementDecision.DENY;
          reason = `blocked_pattern:${pattern}`;
          ruleId = 'tool_policy.blocked';
          break;
        }
      } catch {
        if (pattern && name.includes(pattern)) {
          decision = EnforcementDecision.DENY;
          reason = `blocked_pattern:${pattern}`;
          ruleId = 'tool_policy.blocked';
          break;
        }
      }
    }

    if (decision === EnforcementDecision.ALLOW) {
      for (let i = 0; i < ifcRules.length; i++) {
        const rule = ifcRules[i];
        if (!matchesIfcRule(rule, ctx)) continue;
        if ((rule.deny_tools || []).includes(name)) {
          decision = EnforcementDecision.DENY;
          reason = rule.reason || 'ifc_rule';
          ruleId = `ifc.deny.${i}`;
          break;
        }
        if ((rule.require_approval_tools || []).includes(name)) {
          decision = EnforcementDecision.REQUIRE_APPROVAL;
          reason = rule.reason || 'ifc_rule';
          ruleId = `ifc.approval.${i}`;
          break;
        }
      }
    }

    if (decision === EnforcementDecision.ALLOW && ctx.hasUntrusted && sensitive.has(name)) {
      decision = EnforcementDecision.DENY;
      reason = 'sensitive_tool_with_untrusted_in_scope';
      ruleId = 'ifc.sensitive_default';
    }

    if (decision === EnforcementDecision.ALLOW) {
      const tp = decideToolPolicy(toolPolicy, name);
      if (tp.decision === 'deny') {
        decision = EnforcementDecision.DENY;
        reason = tp.reason;
        ruleId = 'tool_policy.deny';
      } else if (tp.decision === 'require_approval') {
        decision = EnforcementDecision.REQUIRE_APPROVAL;
        reason = tp.reason;
        ruleId = 'tool_policy.approval';
      }
    }

    if (mode === 'development' && decision === EnforcementDecision.DENY) {
      decision = EnforcementDecision.WARN;
    }

    results.push({
      decision,
      tool_name: name,
      reason,
      rule_id: ruleId,
      elapsed_ms: Date.now() - start,
      labels_in_scope: ctx.labelsInScope.toDict()
    });
  }

  return results;
}

module.exports = {
  EnforcementDecision,
  ExecutionContext,
  evaluateTrace,
  decideToolPolicy,
  AuditChain
};
